src/sdk/namespace_fs.js

```diff
@@ -1390,35 +1393,82 @@ class NamespaceFS {
-            const upload_path = path.join(params.mpu_path, `part-${params.num}`);
-            // Will get populated in _upload_stream with the MD5 (if MD5 calculation is enabled)
-            target_file = await this._open_file(fs_context, upload_path, open_mode);
-            const upload_params = { fs_context, params, object_sdk, upload_path, open_mode, target_file, file_path: upload_path };
 
-            await buffers_pool_sem.surround_count(config.NSFS_BUF_SIZE, async () => this._upload_stream(upload_params));
+            const md_upload_path = this._get_part_md_path(params);
+            part_md_file = await this._open_file(fs_context, md_upload_path, md_open_mode);
+            let md_upload_params = {
+                fs_context, params, object_sdk, upload_path: md_upload_path, open_mode: md_open_mode,
+                target_file: part_md_file, file_path: md_upload_path
+            };
 
-            const upload_info = await this._finish_upload(upload_params);
+            let upload_res;
+            const pre_known_size = (params.size >= 0);
+            if (!pre_known_size) {
+                // 2.1
+                upload_res = await buffers_pool_sem.surround_count(config.NSFS_BUF_SIZE, async () => this._upload_stream(md_upload_params));
+                params.size = upload_res.total_bytes;
+            }
+            const offset = params.size * (params.num - 1);
+            const upload_path = this._get_part_data_path(params);
+            dbg.log1(`NamespaceFS: upload_multipart, data path=${upload_path} offset=${offset} data_open_mode=${data_open_mode}`);
+
+            target_file = await this._open_file(fs_context, upload_path, data_open_mode);
+            const data_upload_params = {
+                fs_context, params, object_sdk, upload_path, data_open_mode, target_file,
+                file_path: upload_path, offset
+            };
+
+            if (pre_known_size) {
+                upload_res = await buffers_pool_sem.surround_count(config.NSFS_BUF_SIZE,
+                    async () => this._upload_stream(data_upload_params));
+            } else {
+                // 2.3 if size was not pre known, copy data from part_md_file to by_size file
+                await native_fs_utils.copy_bytes(buffers_pool, fs_context, part_md_file, target_file, params.size, offset);
+            }
+
+            md_upload_params = { ...md_upload_params, offset, digest: upload_res.digest };
+            const upload_info = await this._finish_upload(md_upload_params);
             return upload_info;
         } catch (err) {
             this.run_update_issues_report(object_sdk, err);
             throw this._translate_object_error_codes(err);
         } finally {
             try {
-                if (target_file) await target_file.close(fs_context);
+                if (target_file) target_file.close(fs_context);
+                if (part_md_file) part_md_file.close(fs_context);
```

Comment:
need to await each one individually. I think its time we have a small util function `native_fs_utils.finally_close_files(list_of_files)` that takes a list of files and calls close() on all (maybe with P.map with some small concurrency like 4) and catches the errors on each one individually to print to the log, but it should not throw so that we can call it in finally.